File: src/ChangePointGA/src/model/escorpio17.c
/*
 * escorpio17.c
 *
 *  Created on: Nov 24, 2017
 *      Author: davide
 */

#include "simulation.h"

float getPower (Simulation_ptr simulation, float Ftraction, float Speed, float dt) {
	float MotorPd = 0;
	float ret;
	if (Ftraction>=0)
	{
		// Wheel Model
		float WheelSpeed  = Speed*2/WheelDiameter;
		float WheelTorque = Ftraction*WheelDiameter/2;
		//************

		// Transmission Model
		float MotorSpeed = WheelSpeed/TransmissionRatio;
		float MotorTorque  = WheelTorque*TransmissionRatio*(1/TransmissionEff);
		//************

		// Motor Model
		float MotorVoltage = (MotorSpeed+(MotorSTG*MotorTorque))/MotorSC;
		float MotorCurrent;
		float MotDeltaR = MotorR *( ( (1+simulation->Twindings*alpha_cu) / (1+25.0*alpha_cu) ) -1 );
		if(MotorTorque>0)
		{
			MotorCurrent = (MotorMattrito+MotorTorque)/MotorTC;
			MotorVoltage += MotDeltaR*MotorCurrent*MotorCurrent;
		}
		else
		{
			MotorCurrent = 0;
		}
		// ************

		MotorPd = MotorVoltage*MotorCurrent - MotorSpeed*MotorTorque;

		// Driver Model
#ifdef POWERTRAIN_FLAT
		ret  = Ftraction*Speed/ POWERTRAIN_FLAT_EFF;
#endif

#ifndef POWERTRAIN_FLAT
		float eff = MotorVoltage*MotorCurrent / (MotorVoltage*MotorCurrent + DRV_Pq + DRV_req*MotorCurrent*MotorCurrent + DRV_ki*MotorCurrent + DRV_kv*MotorVoltage);
		if (eff < 0.6) eff = 0.6;
		if (eff > 0.99) eff = 0.99;
		ret = (MotorVoltage*MotorCurrent)/eff;
#endif
	}
	else // Ftraction<0
	{
		// Futura implementazione frenata rigenerativa?
		//ret = 0;

		// Wheel Model
		float WheelSpeed  = Speed*2/WheelDiameter;
		float WheelTorque = Ftraction*WheelDiameter/2;
		//************

		// Transmission Model
		float MotorSpeed = WheelSpeed/TransmissionRatio;
		float MotorTorque  = WheelTorque*TransmissionRatio*(1/TransmissionEff);
		//************

		// Motor Model
		float MotorVoltage = (MotorSpeed+(MotorSTG*MotorTorque))/MotorSC;
		float MotorCurrent;
		float MotDeltaR = MotorR *( ( (1+simulation->Twindings*alpha_cu) / (1+25.0*alpha_cu) ) -1 );
		MotorCurrent = (MotorMattrito+MotorTorque)/MotorTC;
		MotorVoltage += MotDeltaR*MotorCurrent*MotorCurrent;

		MotorPd = MotorVoltage*MotorCurrent - MotorSpeed*MotorTorque;

		// Driver Model
#ifdef POWERTRAIN_FLAT
		ret  = Ftraction*Speed/ POWERTRAIN_FLAT_EFF;
#endif

#ifndef POWERTRAIN_FLAT
		float eff = MotorVoltage*MotorCurrent / (MotorVoltage*MotorCurrent + DRV_Pq + DRV_req*MotorCurrent*MotorCurrent + DRV_ki*MotorCurrent + DRV_kv*MotorVoltage);
		if (eff < 0.6) eff = 0.6;
		if (eff > 0.99) eff = 0.99;
		ret = (MotorVoltage*MotorCurrent)/eff;
#endif
	}

	// Aggiornamento Temperatura del motore
	// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	//if (Sequential)
	//{
		float PtoAmbient = (simulation->TmotorCase-Trear) / RthCaseAmbient;
		float PwindtoCase = (simulation->Twindings-simulation->TmotorCase)/RthWindingCase;
		float PtoCase = PwindtoCase-PtoAmbient;
		float PtoWind = MotorPd - PwindtoCase;
		simulation->Twindings += (PtoWind*dt)/CthWinding;
		simulation->TmotorCase += (PtoCase*dt)/CthCase;
	//}

	// Motor swich-on penalty
	//if (Sequential)
	//{
		if (Ftraction >= MotorStartPenaltyThr)
		{
			if (simulation->lastTractiveForce < MotorStartPenaltyThr)
			{
				ret += MotorStartPenalty/dt;
			}
		}
		simulation->lastTractiveForce = Ftraction;
	//}

	return ret;
}


float getForcebearings(){
	return BearingCr * VehicleMass * gravity;
}

float getForceSlope(float TrackSlope){
	return VehicleMass * gravity * sin_f(TrackSlope);
}

float getForceAero(float Speed, float TrackRadiusInv){
	float Faero = 0.5 * AirDensity * Speed * Speed * FrontalArea * VehicleCx;

	float Yaw = (25 / 0.10) * TrackRadiusInv;  // Deg

	float YawFactor = 3E-17*Yaw*Yaw*Yaw*Yaw*Yaw + 2E-05*Yaw*Yaw*Yaw*Yaw + 5E-14*Yaw*Yaw*Yaw + 0.0037*Yaw*Yaw + 2E-11*Yaw + 1;

	return Faero * YawFactor;
}

float getForceTyres(float Speed, float TrackRadiusInv){
	float TractiveForce;

#ifndef FAST_SIM
	// Nessuna differenza sul verso della sterzata
	float TrackRadiusInvSign = (TrackRadiusInv>0) ? 1.0 : -1.0;
	TrackRadiusInv = fabs(TrackRadiusInv);
	// Tricycle model
	// ripartizione dei pesi
	float Fzr = VehicleMass*gravity*(1 - VehicleXg / VehicleWheelBase);
	//FRAD_r = Fzr;
	float Fzi = VehicleMass*(gravity*VehicleXg / VehicleWheelBase / 2 - VehicleZg*Speed*Speed / VehicleTrack*TrackRadiusInv);
	//FRAD_fi = Fzi;
	if (Fzi<0) Fzi = 0;
	float Fzo = VehicleMass*(gravity*VehicleXg / VehicleWheelBase / 2 + VehicleZg*Speed*Speed / VehicleTrack*TrackRadiusInv);
	//FRAD_fo = Fzo;
	if (Fzo<0) Fzo = 0;
	// caratteristiche delle gomme !!!!!!!!!!!!
	// The magic formula
	float Car = (a30 + a31*TyrePressure)*sin_f(2 * atan_f(Fzr / 1000 / (a40 + a41*TyrePressure))) * 180 / 3.14;
	float Cai = (a30 + a31*TyrePressure)*sin_f(2 * atan_f(Fzi / 1000 / (a40 + a41*TyrePressure))) * 180 / 3.14;
	float Cao = (a30 + a31*TyrePressure)*sin_f(2 * atan_f(Fzo / 1000 / (a40 + a41*TyrePressure))) * 180 / 3.14;
	float Crr = TyresCr * (-0.0156*pow(TyrePressure, 3) + 0.289*pow(TyrePressure, 2) - 1.803*TyrePressure + 4.587);
	float Cri = Crr;
	float Cro = Crr;
	// Sterzo anteriore di Ackerman
	float angleo, anglei;
	if (TrackRadiusInv != 0)
	{
		angleo = atan_f(VehicleWheelBase / (1 / TrackRadiusInv + VehicleTrack / 2));
		anglei = atan_f(VehicleWheelBase / (1 / TrackRadiusInv - VehicleTrack / 2));
	}
	else angleo = anglei = 0;
	//angleo = anglei; // Sterzo parallelo
	// Geometria ipotetica: Camber: lineare 0-10 gradi tra 0-0.1 trackradiusinv
	float Cambo = TrackRadiusInv*10.0 / 0.1*3.14 / 180;
	float Cambi = Cambo = 0;

	// Toe in angle
	angleo += TyresFrontToeIn / 180.0*3.14;
	anglei -= TyresFrontToeIn / 180.0*3.14;
	// Angolo fuori asse
	if (TrackRadiusInvSign < 0)
	{
		angleo -= TyresFrontMisAlignement / 180.0*3.14;
		anglei -= TyresFrontMisAlignement / 180.0*3.14;
	}
	else
	{
		angleo += TyresFrontMisAlignement / 180.0*3.14;
		anglei += TyresFrontMisAlignement / 180.0*3.14;

	}
	// Camber angle
	Cambo += TyresFrontCamber / 180.0*3.14;
	Cambi -= TyresFrontCamber / 180.0*3.14;

	// forza longitudinale
	float Fxr = Crr*Fzr;
	float Fxi = Cri*Fzi*(1 + 0.0005*pow((Cambi*3.14 / 180), 3) - 0.01*pow((Cambi*3.14 / 180), 2) + 0.076*(Cambi*3.14 / 180));
	float Fxo = Cro*Fzo*(1 + 0.0005*pow((Cambo*3.14 / 180), 3) - 0.01*pow((Cambo*3.14 / 180), 2) + 0.076*(Cambo*3.14 / 180));
	// forza centrifuga
	float Fc = Speed*Speed*TrackRadiusInv*VehicleMass;


	// Ricerca iterativa
	// Determino iterativamente il miglior angolo beta che individua la posizione del centro di rotazione
	//float betamax = asin_f(VehicleXg*TrackRadiusInv);

	float Slipo, Slipi, Slipr, e;
	float Fyr, Fyi, Fyo;

	float y, dy, dfyr, dfyo, dfyi;
	float beta = 0;


	//printf("betamax %f\n", betamax);
	float sin_angleo = sin_f(angleo);
	float cos_angleo = cos_f(angleo);
	float sin_anglei = sin_f(anglei);
	float cos_anglei = cos_f(anglei);
	float sin_beta,cos_beta;


	for (int it = 0; it < 5; it++) {
		sin_beta = sin_f(beta);
		cos_beta = cos_f(beta);
		//Slip angles
		if (TrackRadiusInv > 0.0000001 || TrackRadiusInv < -0.0000001) {
			e = VehicleXg - sin_beta / TrackRadiusInv;
			Slipr = atan_f(e*TrackRadiusInv / cos_beta);
			Slipi = anglei - atan_f((VehicleWheelBase - e) / (1 / TrackRadiusInv*cos_beta - VehicleTrack / 2));
			Slipo = angleo - atan_f((VehicleWheelBase - e) / (1 / TrackRadiusInv*cos_beta + VehicleTrack / 2));
		}
		else {
			e = 0;
			Slipr = Slipi = Slipo = 0;
		}
		//Forze trasversali
		/*FLAT_r = */Fyr = Car*Slipr;
		/*FLAT_fi = */Fyi = Cai*Slipi + Cambi*Cai / 10; //!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
		/*FLAT_fo = */Fyo = Cao*Slipo + Cambo*Cao / 10;

		// Equazione dell'equilibrio lungo l'asse trasversale
		y = Fc*cos_beta + Fxo*sin_angleo + Fxi*sin_anglei - Fyr - Fyo*cos_angleo - Fyi*cos_anglei;
		dy = 1;

		if (TrackRadiusInv > 0.0000001 || TrackRadiusInv < -0.0000001) {
			dy = -Fc * sin_beta;

			/*
			dfyr = Car * 1 / (1 + pow(e * TrackRadiusInv / cos_f(beta), 2)) * (-1 + e * TrackRadiusInv * sin_f(beta) / pow(cos_f(beta), 2));

			dfyo = Cao * cos_f(angleo) * (-1 / (1 + pow((VehicleWheelBase - e) / (cos_f(beta) / TrackRadiusInv + VehicleTrack / 2), 2)))
				* ((pow(cos_f(beta) / TrackRadiusInv, 2) + (VehicleTrack * cos_f(beta) + 2 * (VehicleWheelBase - e) * sin_f(beta) / (2 * TrackRadiusInv))) / (pow(cos_f(beta) / TrackRadiusInv + VehicleTrack / 2, 2)));

			dfyi = Cai * cos_f(anglei) * (-1 / (1 + pow((VehicleWheelBase - e) / (cos_f(beta) / TrackRadiusInv - VehicleTrack / 2), 2)))
				* ((pow(cos_f(beta) / TrackRadiusInv, 2) - (VehicleTrack * cos_f(beta) + 2 * (VehicleWheelBase - e) * sin_f(beta) / (2 * TrackRadiusInv))) / (pow(cos_f(beta) / TrackRadiusInv - VehicleTrack / 2, 2)));
			*/

			dfyr = Car * (-1 + e * TrackRadiusInv * sin_beta / pow(cos_beta, 2)) / (1 + pow(e * TrackRadiusInv / cos_beta, 2)) ;

			dfyo = -Cao * cos_angleo *
				(pow(cos_beta / TrackRadiusInv, 2) + (VehicleTrack * cos_beta + 2 * (VehicleWheelBase - e) * sin_beta / (2 * TrackRadiusInv))) /
				(pow(VehicleWheelBase - e, 2) + (pow(cos_beta / TrackRadiusInv + VehicleTrack / 2, 2)));

			dfyi = -Cai * cos_f(anglei) *
				(pow(cos_beta / TrackRadiusInv, 2) - (VehicleTrack * cos_beta + 2 * (VehicleWheelBase - e) * sin_beta / (2 * TrackRadiusInv))) /
				(pow(VehicleWheelBase - e, 2) + (pow(cos_beta / TrackRadiusInv - VehicleTrack / 2, 2)));

			dy -= (dfyr + dfyo + dfyi);
		}

		//printf("%.10f  %.10f   %.10f \n", beta, y, dy);
		beta = beta - y / dy;

	}

	//float eq = Fc*cos_f(beta) + Fxo*sin_f(angleo) + Fxi*sin_f(anglei) - (Fyr + Fyo*cos_f(angleo) + Fyi*cos_f(anglei));


	//printf("beta2 %f  eq %f \n", beta, eq);

	// Debug
	//printf("Speed: %f  e %f\n", Speed, e);

	// Ho finalmente trovato il beta giusto

	TractiveForce = Fxr + Fxo*cos_angleo + Fxi*cos_anglei + Fyo*sin_angleo + Fyi*sin_anglei - Fc*sin_f(beta);
	// Introduce now cornering factor correction
	float straightTractiveForce = Fzi*Cri + Fzo*Cro + Fzr*Crr;

	TractiveForce = TractiveForce + (TractiveForce - straightTractiveForce)*(TyresCorneringCorrFactor - 1) * 2;
	//float TractiveForceb; // Tractive perpendicolare a Fc
	//TractiveForceb = Fxr*cos_f(beta) + Fxo*cos_f(angleo-beta) + Fxi*cos_f(anglei-beta)+ Fyo*sin_f(anglei-beta) + Fyi*sin_f(anglei-beta) - Fyr*sin_f(beta);

	/*
	//if (FrontWheelMaxFy<Fyo*cos_f(angleo)) FrontWheelMaxFy = Fyo*cos_f(angleo);
	//if (FrontWheelMaxFy<Fyi*cos_f(anglei)) FrontWheelMaxFy = Fyi*cos_f(anglei);

	//if (FrontWheelMaxSlip<Slipi) FrontWheelMaxSlip = Slipi;
	//if (FrontWheelMaxSlip<Slipo) FrontWheelMaxSlip = Slipo;
	 */
#else
	// Approssimazione veloce del modello triciclo
	// drag = a + b * (radinv(m))^3/2 * (speed(km/h)+5)^4
	float temp = Speed*3.6 + 5;

	//TractiveForce = 2.368 + 0.0005064 * pow((TrackRadiusInv), 1.5F) *temp*temp*temp*temp;// pow(Speed*3.6+5, 4);
	//TractiveForce = 1973*TyresCr + 0.0005064 * pow((TrackRadiusInv), 1.5F) *temp*temp*temp*temp;// pow(Speed*3.6+5, 4);
	//Dopo correzione 2016:
	float Cr = TyresCr * (-0.0156*pow(TyrePressure, 3) + 0.289*pow(TyrePressure, 2) - 1.803*TyrePressure + 4.587);
	TractiveForce = 0.5342 / 0.0008*Cr + 0.0004782 / 0.0008*Cr * pow((TrackRadiusInv), 1.5F)*TyresCorneringCorrFactor*temp*temp*temp*temp;// pow(Speed*3.6+5, 4);
	// Scale to mass:
	TractiveForce = TractiveForce / 98.0*VehicleMass;
#endif
	// Aero Drag on tyres    !!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	//TractiveForceb += (0.0153*Speed*Speed - 0.0124*Speed)*0.5;
	TractiveForce += (0.00765*Speed*Speed - 0.0062*Speed);

	return TractiveForce;
}


float getForceResistent(float Speed, float TrackSlope, float TrackRadiusInv){
	return getForcebearings() +
			getForceSlope(TrackSlope) +
			getForceAero(Speed, TrackRadiusInv) +
			getForceTyres(Speed, TrackRadiusInv);
}

float getMaxForceTraction(float Speed){
	// Wheel Model
	float WheelSpeed  = Speed*2/WheelDiameter;
	// Transmission Model
	float MotorSpeed = WheelSpeed/TransmissionRatio;
	// Motor Model
	float MotorTorque = ((BatteryVoltage*MotorSC)-MotorSpeed)/MotorSTG;
	// Transmission Model
	float WheelTorque = MotorTorque/TransmissionRatio*TransmissionEff;
	//************
	// Wheel Model
	float Ftraction = WheelTorque/(WheelDiameter/2);
	// Theorical max Ftraction
	// Pratical: limited to 400W, <60N
	if ((Speed*Ftraction)>MaxMotorPower)
		Ftraction = MaxMotorPower/Speed;
	if (Ftraction>MaxWheelForce)
		Ftraction = MaxWheelForce;
	if (Ftraction<0)
		Ftraction = 0;
	return Ftraction;
}

float getForceTraction(float Speed, int map){
	if(map == 0){
		return 0;
	}

	return min(getMapForce(&maps[map - 1], Speed), getMaxForceTraction(Speed));
}

File: src/tfmodel/escorpio17.py

import tensorflow as tf
import numpy as np

#Mass
GRAVITY = tf.constant(9.81, tf.float32, name="GRAVITY")
DRIVER_MASS = tf.constant(56.0, tf.float32, name="DRIVER_MASS")
BARE_BONE_MASS = tf.constant(27.0, tf.float32, name="BARE_BONE_MASS")
MASS = tf.add(DRIVER_MASS, BARE_BONE_MASS, name="MASS")
INERTIAL_MASS = tf.add(MASS, 4.0)

#Bearings
BEARING_CR = tf.constant(0.00008, tf.float32, name="BEARING_CR")

def get_f_bearings():
    return tf.multiply(BEARING_CR, tf.multiply(MASS, GRAVITY), name="f_bearings")
    

def get_f_slope(slope):
	return tf.multiply(tf.multiply(MASS, GRAVITY), tf.sin(slope), name="f_slope")

#Aerodynamic
T_AMB = tf.constant(273.0 + 25.0, tf.float32, name="T_AMB") # Kelvin + degree
AIR_DENSITY = tf.divide(101325.0/287.058, T_AMB, name="AIR_DENSITY")

CX = tf.constant(0.12, tf.float32, name="CX")
FRONTAL_AREA = tf.constant(0.35, tf.float32, name="FRONTAL_AREA")

def get_f_aerodynamic(v, radius):
    yaw = tf.multiply(25.0 / 0.10, radius)
    yaw_factor = tf.add(tf.multiply(3E-17, tf.pow(yaw, 5)), tf.multiply(2E-05, tf.pow(yaw, 4)))
    yaw_factor = tf.add(yaw_factor, tf.add(tf.multiply(5E-14, tf.pow(yaw, 3)), tf.multiply(0.0037, tf.pow(yaw, 2))))
    yaw_factor = tf.add(yaw_factor, tf.add(tf.multiply(2E-11, yaw), 1.0))

    f_aerodynamic = tf.multiply(0.5, tf.multiply(tf.multiply(AIR_DENSITY, tf.pow(v, 2)), tf.multiply(FRONTAL_AREA, CX)))

    return tf.multiply(f_aerodynamic, yaw_factor, name="f_aerodynamic")


#Tyres
WHEEL_BASE = tf.constant(1.60, tf.float32, name="WHEEL_BASE")       # m (passo)
VEHICLE_XG = tf.constant(0.90, tf.float32, name="VEHICLE_XG")       # m (distanza centro di massa - ruota posteriore)
VEHICLE_ZG = tf.constant(0.52, tf.float32, name="VEHICLE_ZG")       # m (distanza centro di massa - piano stradale)
VEHICLE_TRACK = tf.constant(0.52, tf.float32, name="VEHICLE_TRACK") # m (carreggiata)
TYRE_PRESSURE = tf.constant(6.00, tf.float32, name="VEHICLE_XG")    # bar (London pressure) 5.00; // bar (!!!! tra 4 e 6.5 bar)
TYRE_CR = tf.constant(0.0012, tf.float32, name="TYRE_CR")           # (Michelin official 4bar: 0.001)   // Cr a 4bar
TYRE_CORNER_CORR_FACTOR = tf.constant(1.35, tf.float32, name="TYRE_CORNER_CORR_FACTOR")  

TYRE_FRONT_CAMBER = tf.constant(0.1, tf.float32, name="TYRE_FRONT_CAMBER")  # deg (camber ruote anteriori)
TYRE_FRONT_TOE_IN = tf.constant(0.0, tf.float32, name="TYRE_FRONT_TOE_IN")  # deg (toe in ruote anteriori)
TYRE_FRONT_MISALIGNMENT = tf.constant(0.0, tf.float32, name="TYRE_FRONT_MISALIGNMENT")  #  deg (non-allineamento tra ruote anteriori e posteriori)

A30 = tf.constant(57.806, tf.float32, name="A30")      
A31 = tf.constant(15.101, tf.float32, name="A31")       
A40 = tf.constant(-0.082, tf.float32, name="A40")       
A41 = tf.constant(0.186, tf.float32, name="A41")       

def get_f_eq_components(v, radius):
    # Removed sign

    fzr = tf.multiply(MASS, tf.multiply(GRAVITY, tf.subtract(1.0, tf.divide(VEHICLE_XG, WHEEL_BASE))))

    fzi = tf.subtract(
        tf.multiply(GRAVITY, tf.divide(tf.divide(VEHICLE_XG, WHEEL_BASE), 2.0)),
        tf.multiply(tf.divide(tf.multiply(VEHICLE_ZG, tf.pow(v, 2.0)), VEHICLE_TRACK), radius)
    )
    fzi = tf.maximum(0.0, tf.multiply(MASS, fzi))

    fzo = tf.add(
        tf.multiply(GRAVITY, tf.divide(tf.divide(VEHICLE_XG, WHEEL_BASE), 2.0)),
        tf.multiply(tf.divide(tf.multiply(VEHICLE_ZG, tf.pow(v, 2.0)), VEHICLE_TRACK), radius)
    )
    fzo = tf.maximum(0.0, tf.multiply(MASS, fzo))

    car = tf.multiply(
        tf.add(A30, tf.multiply(A31, TYRE_PRESSURE)),
        tf.multiply(
            tf.sin(
                tf.multiply(2.0, tf.atan(
                    tf.divide(tf.divide(fzr, 1000.0), 
                    tf.add(A40, tf.multiply(A41, TYRE_PRESSURE)))
                ))
            ), 
            180.0 / np.pi
        )
    )

    cai = tf.multiply(
        tf.add(A30, tf.multiply(A31, TYRE_PRESSURE)),
        tf.multiply(
            tf.sin(
                tf.multiply(2.0, tf.atan(
                    tf.divide(tf.divide(fzi, 1000.0), 
                    tf.add(A40, tf.multiply(A41, TYRE_PRESSURE)))
                ))
            ), 
            180.0 / np.pi
        )
    )

    cao = tf.multiply(
        tf.add(A30, tf.multiply(A31, TYRE_PRESSURE)),
        tf.multiply(
            tf.sin(
                tf.multiply(2.0, tf.atan(
                    tf.divide(
                        tf.divide(fzo, 1000.0),
                        tf.add(A40, tf.multiply(A41, TYRE_PRESSURE))
                    )
                ))
            ), 
            180.0 / np.pi
        )
    )

    crr = tf.multiply(TYRE_CR,
        tf.add(
            tf.add(
                tf.multiply(-0.0156, tf.pow(TYRE_PRESSURE, 3.0)),
                tf.multiply(0.289, tf.pow(TYRE_PRESSURE, 2.0))
            ),
            tf.add(
                tf.multiply(-1.803, TYRE_PRESSURE),
                4.587
            )
        )
    )

    cri = cro = crr

    # Removed if
    angleo = tf.atan(
        tf.divide(
            WHEEL_BASE,
            tf.add(
                tf.divide(1.0, radius),
                tf.divide(VEHICLE_TRACK, 2.0)
            )
        )
    )

    anglei = tf.atan(
        tf.divide(
            WHEEL_BASE,
            tf.subtract(
                tf.divide(1.0, radius),
                tf.divide(VEHICLE_TRACK, 2.0)
            )
        )
    )

    angleo = tf.add(angleo, tf.divide(TYRE_FRONT_TOE_IN, 180.0 * np.pi))
    anglei = tf.subtract(anglei, tf.divide(TYRE_FRONT_TOE_IN, 180.0 * np.pi))

    # Removed if
    angleo = tf.add(angleo, tf.divide(TYRE_FRONT_MISALIGNMENT, 180.0 * np.pi))
    anglei = tf.add(anglei, tf.divide(TYRE_FRONT_MISALIGNMENT, 180.0 * np.pi))
   
    cambo = tf.multiply(tf.divide(TYRE_FRONT_CAMBER, 180.0), np.pi)
    cambi = tf.multiply(tf.divide(TYRE_FRONT_CAMBER, -180.0), np.pi)

    fxr = tf.multiply(crr, fzr)

    fxi = tf.multiply(
        tf.multiply(cri, fzi),
        tf.add(
            tf.add(
                1.0,
                tf.multiply(0.0005, tf.pow(tf.multiply(cambi, np.pi * 180.0), 3.0))
            ),
            tf.add(
                tf.multiply(-0.01, tf.pow(tf.multiply(cambi, np.pi * 180.0), 2.0)),
                tf.multiply(0.076, tf.multiply(cambi, np.pi * 180.0))
            )
        )
    )
    
    fxo = tf.multiply(
        tf.multiply(cro, fzo),
        tf.add(
            tf.add(
                1.0,
                tf.multiply(0.0005, tf.pow(tf.multiply(cambo, np.pi * 180.0), 3.0))
            ),
            tf.add(
                tf.multiply(-0.01, tf.pow(tf.multiply(cambo, np.pi * 180.0), 2.0)),
                tf.multiply(0.076, tf.multiply(cambo, np.pi * 180.0))
            )
        )
    )
    
    fc = tf.multiply(tf.pow(v, 2.0), tf.multiply(radius, MASS))

    return angleo, anglei, cambo, cambi, car, cao, cai, fxr, fxo, fxi, fc

def get_f_eq(v, radius, beta, angleo, anglei, cambo, cambi, car, cao, cai, fc, fxr, fxo, fxi):
    e = tf.cond(tf.equal(radius, 0.0), 
        true_fn=lambda: tf.zeros(beta.shape), 
        false_fn=lambda: tf.subtract(VEHICLE_XG, tf.divide(tf.sin(beta), radius))
    ) 
    
    slipr = tf.cond(tf.equal(radius, 0.0), 
        true_fn=lambda: tf.zeros(beta.shape), 
        false_fn=lambda: tf.atan(tf.divide(tf.multiply(e, radius), tf.cos(beta)))
    )

    slipi = tf.cond(tf.equal(radius, 0.0), 
        true_fn=lambda: tf.zeros(beta.shape), 
        false_fn=lambda: tf.subtract(
            anglei,
            tf.atan(
                tf.divide(
                    tf.subtract(WHEEL_BASE, e), 
                    tf.subtract(
                        tf.divide(tf.cos(beta), radius), 
                        tf.divide(VEHICLE_TRACK, 2.0)
                    ) 
                )
            )
        )
    )

    slipo = tf.cond(tf.equal(radius, 0.0), 
        true_fn=lambda: tf.zeros(beta.shape), 
        false_fn=lambda: tf.subtract(
            angleo,
            tf.atan(
                tf.divide(
                    tf.subtract(WHEEL_BASE, e), 
                    tf.add(
                        tf.divide(tf.cos(beta), radius), 
                        tf.divide(VEHICLE_TRACK, 2.0)
                    ) 
                )
            )
        )
    )
   
    fyr = tf.multiply(car, slipr)
    fyi = tf.add(tf.multiply(cai, slipi), tf.divide(tf.multiply(cambi, cai), 10.0))          
    fyo = tf.add(tf.multiply(cao, slipo), tf.divide(tf.multiply(cambo, cao), 10.0))          

    eq = tf.subtract(
            tf.add(
                tf.add(
                    tf.multiply(fc, tf.cos(beta)),
                    tf.multiply(fxo, tf.sin(angleo))
                ),
                tf.subtract(
                    tf.multiply(fxi, tf.sin(anglei)),
                    fyr
                )
            ),
            tf.add(
                tf.multiply(fyo, tf.cos(angleo)),
                tf.multiply(fyi, tf.cos(anglei)) 
            )
        )

    return eq, beta, fyo, fyi

def get_optimal_beta(v, radius, angleo, anglei, cambo, cambi, car, cao, cai, fc, fxr, fxo, fxi):
    opt_beta = tf.zeros(v.shape)

    eq = get_f_eq(v, radius, opt_beta, angleo, anglei, cambo, cambi, car, cao, cai, fc, fxr, fxo, fxi)[0]
    grad = tf.gradients(eq, opt_beta)[0]
    
    for i in range(0, 5):
        eq = get_f_eq(v, radius, opt_beta, angleo, anglei, cambo, cambi, car, cao, cai, fc, fxr, fxo, fxi)[0]
        opt_beta = tf.subtract(opt_beta, 
            tf.cond(
                tf.equal(radius, 0.0),
                true_fn= lambda: tf.zeros(v.shape),
                false_fn= lambda: tf.divide(eq, grad)
            )
        )

    return opt_beta
   

def get_f_tyres(v, radius):
    angleo, anglei, cambo, cambi, car, cao, cai, fxr, fxo, fxi, fc = get_f_eq_components(v, radius)
   
    #opt_beta = tf.zeros(v.shape)
    opt_beta = get_optimal_beta(v, radius, angleo, anglei, cambo, cambi, car, cao, cai, fc, fxr, fxo, fxi)

    feq, beta, fyo, fyi = get_f_eq(v, radius, opt_beta, angleo, anglei, cambo, cambi, car, cao, cai, fc, fxr, fxo, fxi)

    f_tractive = tf.add(
        tf.add(
            tf.add(
                fxr,
                tf.multiply(fxo, tf.cos(angleo))
            ),
            tf.add(
                tf.multiply(fxi, tf.cos(anglei)),
                tf.multiply(fyo, tf.sin(angleo))
            )
        ),
        tf.subtract(
            tf.multiply(fyi, tf.sin(anglei)),
            tf.multiply(fc, tf.sin(opt_beta))
        )
    )

    f_tractive = tf.add(
        f_tractive,
        tf.subtract(
            tf.multiply(0.00765, tf.pow(v, 2.0)),
            tf.multiply(0.0062, v)
        )
    )

    return f_tractive


WHEEL_DIAMETER = tf.constant(0.466, tf.float32, name="WHEEL_DIAMETER")
TRANSMISSION_RATEO = tf.constant(1.0 / 32.0, tf.float32, name="TRANSMISSION_RATEO")
TRANSMISSION_EFF = tf.constant(1.0 / 32.0, tf.float32, name="TRANSMISSION_EFF")
MOTOR_SC = tf.constant(25.2, tf.float32, name="MOTOR_SC")
MOTOR_STG = tf.constant(71.93, tf.float32, name="MOTOR_STG")

BATTERY_VOLTAGE = tf.constant(46.0, tf.float32, name="BATTERY_VOLTAGE")

MAX_MOTOR_POWER = tf.constant(500.0, tf.float32, name="MAX_MOTOR_POWER")
MAX_WHEEL_FORCE = tf.constant(65.0, tf.float32, name="MAX_WHEEL_FORCE")

def get_f_traction_max(v):
    wheel_speed = tf.divide(tf.multiply(2.0, v), WHEEL_DIAMETER)

    motor_speed = tf.divide(wheel_speed, TRANSMISSION_RATEO)
	
    motor_torque = tf.divide(tf.subtract(tf.multiply(BATTERY_VOLTAGE, MOTOR_SC), motor_speed), MOTOR_STG) 
    
    wheel_torque = tf.multiply(tf.divide(motor_torque, TRANSMISSION_RATEO), TRANSMISSION_EFF) 
	
    f_traction = tf.divide(wheel_torque, tf.divide(WHEEL_DIAMETER, 2.0))
	
    f_traction = tf.minimum(f_traction, tf.divide(MAX_MOTOR_POWER, v))
    f_traction = tf.minimum(f_traction, MAX_WHEEL_FORCE)
    f_traction = tf.maximum(f_traction, 0.0)

    return f_traction
